src/resolve-protocol.ts

```diff
@@ -42,8 +42,11 @@ const connect = async (proxyUrl: string, options: tls.ConnectionOptions, callbac
 
             request.once('connect', (response, socket, head) => {
                 if (response.statusCode !== 200 || head.length > 0) {
-                    reject(new Error(`Proxy responded with ${response.statusCode} ${response.statusMessage}: ${head.length} bytes`));
+                    reject(new Error(`Proxy responded with ${response.statusCode} ${response.statusMessage}: ${head.length} bytes.
 
+Below is the first 100 bytes of the proxy response body:
+${head.toString('utf8', 0, 100)}
+`));
```

Comment:
Seems weird to include this kind of formatting in the message of an `Error`, but we probably don't have any other options, do we?

Reply:
we could squish it in [`cause`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause), but the intuition (and MDN doc) tells me that the `message` is for human readable errors and `cause` is for structured data / original errors in case of rethrowing.

Reply:
Yeah, sounds reasonable. Still, the newlines freak me out :smile: 

Could we also include the whole response body somewhere in `cause`?

Reply:
multiline error messages are quite common, we use those in several places already, also playwright and puppeteer do the same (and much worse, e.g. 10+ lines error messages, namely those about a missing browser executable). you can attach metadata to the error instance too, but its better to have a custom error class for that so its discoverable. we also use this approach (but we don't add any metadata there, just custom error prototypes).

Reply:
Had to bump `got` to have https://github.com/sindresorhus/got/pull/2327 , which caused the build to break... fixed by the `tsconfig.json` update. Happy to get feedback on that, it fixes the build, but might not be the best way around it.